gmso/formats/gro.py

```diff
@@ -116,77 +120,92 @@ def write_gro(top, filename):
             )
         )
         out_file.write("{:d}\n".format(top.n_sites))
-        for idx, site in enumerate(top.sites):
-            warnings.warn(
-                "Residue information is not currently "
-                "stored or written to GRO files.",
-                NotYetImplementedWarning,
-            )
-            # TODO: assign residues
-            res_id = 1
-            res_name = "X"
-            atom_name = site.name
-            atom_id = idx + 1
-            out_file.write(
-                "{0:5d}{1:5s}{2:5s}{3:5d}{4:8.3f}{5:8.3f}{6:8.3f}\n".format(
-                    res_id,
-                    res_name,
-                    atom_name,
-                    atom_id,
-                    site.position[0].in_units(u.nm).value,
-                    site.position[1].in_units(u.nm).value,
-                    site.position[2].in_units(u.nm).value,
-                )
-            )
-
-        if allclose_units(
-            top.box.angles,
-            u.degree * [90, 90, 90],
-            rtol=1e-5,
-            atol=0.1 * u.degree,
-        ):
-            out_file.write(
-                " {:0.5f} {:0.5f} {:0.5f} \n".format(
-                    top.box.lengths[0].in_units(u.nm).value.round(6),
-                    top.box.lengths[1].in_units(u.nm).value.round(6),
-                    top.box.lengths[2].in_units(u.nm).value.round(6),
-                )
-            )
-        else:
-            # TODO: Work around GROMACS's triclinic limitations #30
-            vectors = top.box.get_vectors()
-            out_file.write(
-                " {:0.5f} {:0.5f} {:0.5f} {:0.5f} {:0.5f} {:0.5f} {:0.5f} {:0.5f} {:0.5f} \n".format(
-                    vectors[0, 0].in_units(u.nm).value.round(6),
-                    vectors[1, 1].in_units(u.nm).value.round(6),
-                    vectors[2, 2].in_units(u.nm).value.round(6),
-                    vectors[0, 1].in_units(u.nm).value.round(6),
-                    vectors[0, 2].in_units(u.nm).value.round(6),
-                    vectors[1, 0].in_units(u.nm).value.round(6),
-                    vectors[1, 2].in_units(u.nm).value.round(6),
-                    vectors[2, 0].in_units(u.nm).value.round(6),
-                    vectors[2, 1].in_units(u.nm).value.round(6),
-                )
-            )
+        out_file.write(_prepare_atoms(top))
+        out_file.write(_prepare_box(top))
 
 
-def _prepare_topology_to_gro(top):
+def _validate_positions(top):
     """Modify topology, as necessary, to fit limitations of the GRO format."""
     if np.min(top.positions) < 0:
         warnings.warn(
             "Topology contains some negative positions. Translating "
             "in order to ensure all coordinates are non-negative."
         )
-
+    min_xyz = np.min(top.positions, axis=0)
+    for i, minimum in enumerate(min_xyz):
+        if minimum < 0.0:
+            for site in top.sites:
+                site.position[i] = site.position[i] - minimum
     return top
```

Comment:
For this method, do we just want to work on a clone of the `top` and not the `top` itself? We might bump into the issue with LAMMPS writer in mbuild if we are modifying the original object.